govcd/entity_client.go

```diff
@@ -0,0 +1,225 @@
+	if err != nil {
+		return fmt.Errorf("error in HTTP DELETE request: %s", err)
+	}
+
+	err = resp.Body.Close()
+	if err != nil {
+		return fmt.Errorf("error closing response body: %s", err)
+	}
+
+	return nil
+}
+
+func (client *Client) performEntityPostPut(httpMethod string, urlRef *url.URL, params url.Values, payload interface{}, additionalHeader map[string]string) (*http.Response, error) {
+	// Marshal payload if we have one
+	body := new(bytes.Buffer)
+	if payload != nil {
+		marshaledJson, err := json.MarshalIndent(payload, "", "  ")
+		if err != nil {
+			return nil, fmt.Errorf("error marshalling JSON data for %s request %s", httpMethod, err)
+		}
+		body = bytes.NewBuffer(marshaledJson)
+	}
+
+	req := client.newEntityRequest(params, httpMethod, urlRef, body, additionalHeader)
+	resp, err := client.Http.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// resp is ignored below because it is the same the one above
+	_, err = checkRespWithErrType(types.BodyTypeJSON, resp, err, &ccitypes.ApiError{})
+	if err != nil {
+		return nil, fmt.Errorf("error in HTTP %s request: %s", httpMethod, err)
+	}
+	return resp, nil
+}
+
+func (client *Client) newEntityRequest(params url.Values, method string, reqUrl *url.URL, body io.Reader, additionalHeader map[string]string) *http.Request {
+	// copy passed in URL ref so that it is not mutated
+	reqUrlCopy := copyUrlRef(reqUrl)
+
+	// Add the params to our URL
+	reqUrlCopy.RawQuery += params.Encode()
+
+	// If the body contains data - try to read all contents for logging and re-create another
+	// io.Reader with all contents to use it down the line
+	var readBody []byte
+	var err error
+	if body != nil {
+		readBody, err = io.ReadAll(body)
+		if err != nil {
+			util.Logger.Printf("[DEBUG - newEntityRequest] error reading body: %s", err)
+		}
+		body = bytes.NewReader(readBody)
+	}
+
+	req, err := http.NewRequest(method, reqUrlCopy.String(), body)
+	if err != nil {
+		util.Logger.Printf("[DEBUG - newEntityRequest] error getting new request: %s", err)
+	}
+
+	if client.VCDAuthHeader != "" && client.VCDToken != "" {
+		// Add the authorization header
+		req.Header.Add(client.VCDAuthHeader, client.VCDToken)
+		// The deprecated authorization token is 32 characters long
+		// The bearer token is 612 characters long
+		if len(client.VCDToken) > 32 {
+			req.Header.Add("Authorization", "bearer "+client.VCDToken)
+			req.Header.Add("X-Vmware-Vcloud-Token-Type", "Bearer")
+		}
+	}
+
+	for k, v := range client.customHeader {
+		for _, v1 := range v {
+			req.Header.Set(k, v1)
+		}
+	}
+	for k, v := range additionalHeader {
+		req.Header.Add(k, v)
+	}
+
+	// Inject JSON mime type if there are no overwrites
+	if req.Header.Get("Content-Type") == "" {
+		req.Header.Add("Content-Type", types.JSONMime)
+	}
+
+	setHttpUserAgent(client.UserAgent, req)
+	// TODO - can we send a traceable request ID somewhere so that responses contain it for better log tracing
+	// setVcloudClientRequestId(client.RequestIdFunc, req)
```

Comment:
it will be useful to send the `X-VMWARE-VCLOUD-CLIENT-REQUEST-ID` trace id

Reply:
Ok! Sending as it is now